src/mlpack/methods/ann/layer/minibatch_discrimination.hpp

```diff
@@ -0,0 +1,190 @@
+  //! Create the MiniBatchDiscrimination object.
+  MiniBatchDiscrimination();
+
+  /**
+   * Create the MiniBatchDiscrimination layer object using the specified
+   * number of units.
+   *
+   * @param inSize The number of input units.
+   * @param outSize The number of output units.
+   * @param features The number of features to compute for each dimension.
+   */
+  MiniBatchDiscrimination(const size_t inSize,
+                          const size_t outSize,
+                          const size_t features);
+
+  /**
+   * Reset the layer parameter.
+   */
+  void Reset();
+
+  /**
+   * Ordinary feed-forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Ordinary feed-backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards through f. Using the results from the
+   * feed-forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Calculate the gradient using the output delta and the input activation.
+   *
+   * @param input The input parameter used for calculating the gradient.
+   * @param error The calculated error.
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(const arma::Mat<eT>&& input,
+                arma::Mat<eT>&& /* error */,
+                arma::Mat<eT>&& gradient);
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored dimensions of weight.
+  size_t A, B, C;
```

Comment:
Can we use a more descriptive name here?

Reply:
I have used the names and dimensions as described in the [paper](https://arxiv.org/pdf/1606.03498.pdf). Thought it would be easier to understand for someone to follow the implementation after reading the paper. 
I could use `rows` and `cols` but that wouldn't be very meaningful because I keep reshaping the matrices.

Reply:
I see, thanks for the clarification.